File: deck/base/code/native/swift/runtime/http.swift
import Foundation

enum http {
    static func request(_ method: String, _ url: String, _ body: String) async -> HttpResponse {
        guard let u = URL(string: url) else { return HttpResponse(status: 0, body: "") }
        var req = URLRequest(url: u)
        req.httpMethod = method
        if !body.isEmpty { req.httpBody = body.data(using: .utf8) }
        do {
            let (data, response) = try await URLSession.shared.data(for: req)
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            return HttpResponse(status: status, body: String(data: data, encoding: .utf8) ?? "")
        } catch {
            return HttpResponse(status: 0, body: "")
        }
    }
}

File: deck/base/code/native/rust/runtime/http.rs
mod http {
    use super::HttpResponse;
    pub async fn request(method: String, url: String, body: String) -> HttpResponse {
        let client = ::reqwest::Client::new();
        let verb = method.parse::<::reqwest::Method>().unwrap_or(::reqwest::Method::GET);
        let mut builder = client.request(verb, &url);
        if !body.is_empty() { builder = builder.body(body); }
        match builder.send().await {
            Ok(response) => {
                let status = response.status().as_u16() as i64;
                let text = response.text().await.unwrap_or_default();
                HttpResponse { status, body: text }
            }
            Err(_) => HttpResponse { status: 0, body: String::new() },
        }
    }
}
